Introduce WithFacingSpriteBody and related With*Animation traits

reaperrr opened the pull request:
Another (large) step towards getting rid of `RenderSimple` and all traits based on it, and instead use `RenderSprites` + `WithSpriteBody`-based traits and additional `With*Animation` traits.

This PR modifies `WithSpriteBody` to be roughly equivalent to `RenderSimple` and introduces `WithFacingSpriteBody` for moving actors with facings, as well as 
1) `WithMoveAnimation` to replace `RenderUnitFlying` special cases and 
2) `WithAttackAnimation` to replace `RenderUnitReload` special case as well as allow the use of `WithFacingSpriteBody` in cases where `WithInfantryBody` would be overkill (example: TS Wolverine Mech).

This should fix the issue with the Wolverine occasionally getting stuck on the last shooting frame as a side-effect.

Once DeployToUpgrade lands, we can probably use this to implement TS deployable units better since With*Body is upgradable.

reaperrr committed:
Added WithFacingSpriteBody trait.

Extracted AutoSelectionSize into its own trait.

RenderUnit -> RenderSprites + WithFacingSpriteBody upgrade rule

Remove RenderSimple dependencies of ThrowsParticle
and WithHarvestAnimation.

reaperrr committed:
Add deprecation notices to RenderUnit and RenderSimple

reaperrr committed:
Added WithAttackAnimation trait

Interacts with WithUnitBody.

Upgrade rule for RenderUnitReload -> RenderSprites + WithUnitBody + WithAttackAnimation

Use WithAttackAnimation for RA V2 launcher
Remove RenderUnitReload

reaperrr committed:
Added WithMoveAnimation trait

RenderUnitFlying -> WithUnitBody + WithMoveAnimation upgrade rule

Make TD Orca use WithMoveAnimation
Remove RenderUnitFlying

reaperrr committed:
TS wolverine changed to use WithFacingSpriteBody

...in conjunction with WithMoveAnimation and WithAttackAnimation,
replacing WithInfantryBody.

reaperrr committed:
Upgrade TD to replace RenderUnit with WithFacingSpriteBody

+RenderSprites + AutoSelectionSize.

reaperrr committed:
Replace RenderUnit in RA mod with WithFacingSpriteBody

+RenderSprites + AutoSelectionSize.

Change MadTank dependency from RenderUnit to WithFacingSpriteBody.

reaperrr committed:
Replace RenderUnit in D2k with WithFacingSpriteBody

+RenderSprites + AutoSelectionSize, where possible.

reaperrr committed:
Upgrade TS to replace RenderUnit with WithFacingSpriteBody

+RenderSprites + AutoSelectionSize.

pchote commented:
You'll need to update a bunch of `Requires<RenderSimpleInfo>` &rarr; `Requires<ISpriteBodyInfo>` to fix the failing tests.

reaperrr commented:
Updated. Sorry for the comment loss, but I did adress everything that was pointed out/reported by Travis (including moving the ```MakeFaceFunc``` stuff to ```WithSpriteBody``` ctor and making DefaultAnimation ```readonly```).

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderSprites.cs:
```diff
@@ -229,8 +229,8 @@ public static string NormalizeSequence(Animation anim, DamageState state, string
-		// Required by RenderSimple
-		protected int2 AutoSelectionSize(Actor self)
+		// Required by RenderSimple, WithSpriteBody and WithInfantryBody
+		public int2 AutoSelectionSize(Actor self)
```
We'll eventually want to move / remove this, but that can wait until after we remove `RenderSimple`.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -0,0 +1,73 @@
+			if (!string.IsNullOrEmpty(prefix))
+				sequence = prefix + sequence;
+			if (sequence != wub.DefaultAnimation.CurrentSequence.Name)
+				wub.PlayCustomAnimationRepeating(self, sequence);
```
Shouldn't this be `wub.DefaultAnimation.ReplaceAnim(sequence)`?

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:
```diff
@@ -9,44 +9,83 @@
 		public readonly string StartSequence = null;
 
-		[Desc("Animation to play when the actor is idle.")]
+		[Desc("Animation to play when the actor is idle/standing.")]
```
Not sure I agree with adding "standing" here.  That is only true for infantry.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:
```diff
@@ -9,44 +9,83 @@
+		public readonly Animation DefaultAnimation;
 
-		public WithSpriteBody(ActorInitializer init, WithSpriteBodyInfo info)
+		public WithSpriteBody(ActorInitializer init, WithSpriteBodyInfo info, Func<int> baseFacing)
```
If a subclass calls this directly then we will ignore any `StartSequence` that may be defined.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:
```diff
@@ -9,44 +9,83 @@
+			return RenderSprites.NormalizeSequence(DefaultAnimation, self.GetDamageState(), sequence);
+		}
+
+		public void PlayCustomAnim(Actor self, string name)
```
Could you please squash this into `PlayCustomAnimation` below?

pchote replied:
Just define a default value for `Action after = null`, add the sequence check (although I think we may prefer it to crash instead, as people really shouldn't be calling this with bogus sequences), and then change the callers to use the other method.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithUnitBody.cs:
```diff
@@ -0,0 +1,26 @@
+	public class WithUnitBody : WithSpriteBody
+	{
+		public WithUnitBody(ActorInitializer init, WithUnitBodyInfo info)
+			: base(init, info) { }
```
With my earlier comment I meant passing the facing function from here.  Otherwise, the `Requires<IFacingInfo>` above doesn't make sense, because this trait doesn't (explicitly) use IFacing.

reaperrr commented:
Rebased and adressed some of the issues that were pointed out.

reaperrr commented:
Updated with @pchote's changes.

About a different name for ```WithUnitBody```, the most descriptive name would probably be ```WithFacingSpriteBody```, but that's pretty long. ```WithFacingBody``` is shorter, but doesn't make clear that it's not for voxels, for example.

penev92 commented:
I like `WithFacingSpriteBody`/`WithRotatingSpriteBody`, as they are pretty self-explanatory. Especially the second one. It's 2015, we have good IDEs, and should not be afraid of long names.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -0,0 +1,73 @@
+			if (!string.IsNullOrEmpty(info.AimSequence) && attack.IsAttacking)
+				sequence = info.AimSequence;
+
+			var prefix = armament.IsReloading && !string.IsNullOrEmpty(info.ReloadPrefix) ? info.ReloadPrefix : "";
```
Please add braces around the condition for readability:
`var prefix = (armament.IsReloading && !string.IsNullOrEmpty(info.ReloadPrefix)) ? info.ReloadPrefix : "";`

penev92 commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1016,6 +1016,75 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+					var rur = node.Value.Nodes.Where(x => x.Key == "RenderUnitReload");
+					if (rur.Any())
+					{
+						rur.Do(x => x.Key = "RenderSprites");
```
Are you sure this works? Because I've tried using `.Do()` to modify collections and it didn't.

reaperrr replied:
*shrugs* My upgrade rules are copied and modified from other, working upgrade rules, mostly @pchote's WithInfantryBody upgrade rules. I did test them once, it worked for the ra V2.

penev92 replied:
Good enough for me.

penev92 commented:
I theory, I love this. It takes care of point 1. (of 5) of the rendering traits plan we discussed.
The code looks good, but I will need to get back to you about the upgrade rules and the YAML changes (damn, there's a lot of them!).
All YAML changes were done by the upgrade rules, right? Or did you do them manually?

reaperrr commented:
> All YAML changes were done by the upgrade rules, right? Or did you do them manually?

~99% were done via upgrade rules, yes. Doing this manually is more risky for this amount of yaml code.
The only manual work is the TS wolverine since it used WithInfantryBody.

reaperrr commented:
Updated. There was a WithUnitBody rename leftover on the ra V2 (not in the end result, but 6d532be renamed to WithUnitBody and efc2564 then renamed WUB to WFSB, now the first commit renames to WFSB right away), fixed that as well.

penev92 commented on mods/ra/maps/allies-01/map.yaml:
```diff
@@ -488,12 +487,17 @@ Rules:
 			MaxWeight: 0
+		AutoSelectionSize:
+		RenderSprites:
+			Image: tran
```
These can probably be removed, as they are inherited from `TRAN.Extraction`. It's worth to check about `RenderSprites` though.

penev92 commented:
OK, so:
 - [ ] Flares and oil derrick husks still work as before.
 - [ ] V2 still works, although the facings are less than perfect (not sure if this is a regression though, iirc they weren't too good before either).
 - [ ] Wolverine works with no noticeable issues.
 - [ ] Visceroids in TS seems to work OK.
 - [ ] Everything in D2k seems to work.

This introduces a lot of unneeded `AutoSelectionSize:` traits. For example on husks. Actually I'm not sure what the point of having that in its own trait is, as it seems to just call one method from `RenderSprites`.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -0,0 +1,73 @@
+		public void Attacking(Actor self, Target target, Armament a, Barrel barrel)
+		{
+			if (!string.IsNullOrEmpty(info.AttackSequence))
+				wfsb.PlayCustomAnimation(self, info.AttackSequence, null);
```
Redundant `, null`.

reaperrr commented:
> This introduces a lot of unneeded AutoSelectionSize: traits. For example on husks. Actually I'm not sure what the point of having that in its own trait is, as it seems to just call one method from RenderSprites.

@pchote asked to move that into its own trait, the method will eventually be moved from ```RenderSprites``` to this trait.

I will clean up the yamls in a follow-up PR, but for the scope of this PR I felt the result of the upgrade rules was sufficient.

reaperrr commented:
Updated.

> V2 still works, although the facings are less than perfect (not sure if this is a regression though, iirc they weren't too good before either).

The ```aim``` art is custom-made and has only 8 facings.

Mailaender commented:
Needs a rebase.

reaperrr commented:
Rebased.

Mailaender commented:
Can't spot any regressions in the related changes as well. :white_check_mark:

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/620340145a5bbc88c6df0849113be4a456f273f5...6121b899dc198dcc76902307465e446a2f1ecf07)

Unit158 commented:
This improves custom animation code greatly. Thanks :)

On 2015-06-07, at 11:18 AM, Matthias Mailänder <notifications@github.com> wrote:

> Changelog
> 
> —
> Reply to this email directly or view it on GitHub.
>
